File: string_calculator/1-python/string_calculator.py
#

class StringCalculator():
    MAX_VALUE = 1000
    SEPARATOR = ","
    def add(self, _string):
        if _string == "": return 0
        ret = 0
        normalized_string = self.normalizeString(_string)
        numbers = normalized_string.split(StringCalculator.SEPARATOR)
        errors = ""
        for nn in numbers:
            if nn[0] == "-":
                if errors != "":
                    errors += "\n"
                errors += "negative numbers not allowed : " + nn
                continue
            val = int(nn)
            if val < StringCalculator.MAX_VALUE:
                ret += val
        if errors != "":
            raise Exception(errors)
        return ret

    def normalizeString(self, _string):
        delimiters = None
        if _string[:2] == "//":
            delimiter, _string  = _string[2:].split("\n", 1)
            if "[" in delimiter:
                delimiters = delimiter[1:-1].split("][")
            else:
                delimiters = [delimiter, ]
        _string = _string.replace("\n", StringCalculator.SEPARATOR)
        if delimiters:
            for delimiter in  delimiters:
                _string = _string.replace(delimiter, StringCalculator.SEPARATOR)
        return _string

#

File: string_calculator/2-java/StringCalculator.java
package stringCalculator;


public class StringCalculator {
	String errors = "";
	
	int add(String numbers) {
		if (numbers.isEmpty() == true) {
			return 0;
		}
		return computeSumOfOperators(numbers);
	}

	private Integer computeSumOfOperators(String numbers) {
		String[] operands = getOperandsAsStringsList(numbers);
		resetErrorMessage();
		int ret = 0;
		for (String operand : operands) {
			ret = checkAndAddValue(ret, operand);
		}
		throwExceptionIfErrorMessage();
		return ret;
	}

	private int checkAndAddValue(int sum, String val) {
		int ret = sum;
		int value = Integer.valueOf(val);
		if (value < 0) {
			addErrorMessage("negatives not allowed : " + val);
		}
		else if (value < 1000) {
			ret += value;
		}
		return ret;
	}

	private void resetErrorMessage() {
		errors = "";
	}
	
	private void addErrorMessage(String val) {
		if (!errors.equals("")) {
			errors = errors + "\n";
		}
		errors = errors + val;
	}

	private void throwExceptionIfErrorMessage() {
		if (errors.isEmpty() == false) {
			throw new RuntimeException(errors);
		}
	}

	private String[] getOperandsAsStringsList(String operands) {
		String normalized_numbers = normalizeOperands(operands);
		return normalized_numbers.split("\n");
	}

	private String normalizeOperands(String operands) {
		String[] separators = getCustomSeparatorList(operands);
		String numbers_only = getOperatorsOnly(operands);
		String ret = numbers_only;
		for (String separator : separators) {
			ret = ret.replace(separator, "\n");
		}
		return ret;
	}

	private boolean hasCustomSeparator(String operands) {
		return operands.length() > 2 && operands.substring(0, 2).equals("//");
	}

	private String getOperatorsOnly(String operands) {
		String ret = operands;
		if (hasCustomSeparator(operands)) {
			ret = operands.substring(operands.indexOf("\n")+1);
		}
		return ret;
	}

	private String[] getCustomSeparatorList(String operands) {
		if (!hasCustomSeparator(operands)) {
			return getSingleSeparatorList(",");
		}
		String[] ret;
		String separator = operands.substring(2, operands.indexOf("\n"));
		if (separator.startsWith("[")) {
			separator = separator.substring(1, separator.length()-1);
			ret = separator.split("\\]\\[");
			return ret;
		}
		return getSingleSeparatorList(separator);
	}

	private String[] getSingleSeparatorList(String separator) {
		String[] ret = new String[1];
		ret[0] = separator;
		return ret;
	}
	
}
